decisions/0002-lazy-route-modules.md

```diff
@@ -0,0 +1,256 @@
+
+### Error Handling
+
+If an error is thrown by `lazy()` we catch that in the same logic as iof the error was thrown by the `action`/`loader` and bubble it to the nearest `errorElement`.
+
+## Consequences
+
+Not so much as a consequence, but more of limitation - we still require the routing tree up front-for the most efficient data-loading. This means that we can't _yet_ support quite the same nested `<Routes>` use-cases as before (particularly with respect to microfrontends), but we have ideas for how to solve tht as an extension of this concept in the future.
+
+Another slightly edge-case concept we discovered is that in DIY SSR applications using `createStaticHandler` and `StaticRouterProvider`, it's possible to server-render a lazy route and send up it's hydration data. But then we may _not_ have those routes loaded in our client-side hydration:
+
+```jsx
+const routes = [{
+  path: '/',
+  lazy: () => import("./route"),
+}]
+let router = createBrowserRouter(routes, {
+  hydrationData: window.__hydrationData,
+});
+
+// ⚠️ At this point, the router has the data but not the route definition!
+
+ReactDOM.hydrateRoot(
+  document.getElementById("app")!,
+  <RouterProvider router={router} fallbackElement={null} />
+);
+```
+
+In the above example, we've server-rendered our `/` route and therefore we _don't_ want to render a `fallbackElement` since we already have the SSR'd content, and the router doesn't need to "initialize" because we've provided the data in `hydrationData`. However, if we're hydrating into a route that includes `lazy`, then we _do_ need to initialize that lazy route.
+
+The real solution for this is to do what Remix does and know your matched routes and preload their modules ahead of time and hydrate with synchronous route definitions. This is a non-trivial process through so it's not expected that every DIY SSR use-case will handle it. Instead, the router will not be initialized until any initially matched lazy routes are loaded, and therefore we need to delay the hydration or our `RouterProvider`. We can do this in one of two ways:
+
+**Option 1 - preemptively load lazy initial matches:**
+
+```jsx
+// Determine if any of the initial routes are lazy
+let lazyMatches = matchRoutes(routes, window.location)?.filter(
+  (m) => m.route.lazy
+);
+
+// Load the lazy matches and update the routes before creating your router
+// so we can hydrate the SSR-rendered content synchronously
+if (lazyMatches && lazyMatches?.length > 0) {
+  await Promise.all(
+    lazyMatches.map(async (m) => {
+      let routeModule = await m.route.lazy!();
+      Object.assign(m.route, { ...routeModule, lazy: undefined });
+    })
+  );
+}
+
+createRouterAndHydrate();
+```
+
+**Option 2 - Wait for router to load lazy initial matches and set `state.initialized=true`:**
```

Comment:
IMO this is a worse option when considering the possibility that a `lazy` call could fail. I'd be tempted to omit it entirely and just present the first option which feels clearer to me. What do you think?

Reply:
Yeah that feels better